packages/testcontainers/src/containers/Hydra/NativeChainContainer.ts

```diff
@@ -0,0 +1,225 @@
+import {
+  GenericContainer,
+  Network,
+  StartedTestContainer
+} from 'testcontainers'
+import fetch from 'cross-fetch'
+import { AbstractStartedContainer } from 'testcontainers/dist/modules/abstract-started-container'
+import { getNetwork, Network as JellyfishNetwork, NetworkName } from '@defichain/jellyfish-network'
+import { JsonRpcClient } from '@defichain/jellyfish-api-jsonrpc'
+
+/**
+ * Mandatory options to start NativeChain with
+ */
+export interface StartOptions {
+  // TODO(fuxingloh): change to cookie based auth soon
+  user: string
+  password: string
+  timeout?: number
+  startFlags?: StartFlags[]
+}
+
+export interface StartFlags {
+  name: string
+  value: number
+}
+
+/**
+ * DeFiChain NativeChain node managed in docker
+ */
+export class NativeChainContainer extends GenericContainer {
+  /**
+   * @param {string} image docker image name
+   */
+  constructor (
+    image: string = NativeChainContainer.image
+  ) {
+    super(image)
+  }
+
+  static get image (): string {
+    if (process?.env?.DEFICHAIN_DOCKER_IMAGE !== undefined) {
+      return process.env.DEFICHAIN_DOCKER_IMAGE
+    }
+    return 'defi/defichain:HEAD-02ef6a1b3'
+  }
+
+  public static readonly DefaultStartOptions = {
+    user: 'testcontainers-user',
+    password: 'testcontainers-password'
+  }
+
+  protected startOptions: StartOptions = NativeChainContainer.DefaultStartOptions
+
+  public withStartOptions (startOptions: StartOptions): this {
+    this.startOptions = startOptions
+    return this
+  }
+
+  protected networkConfig: JellyfishNetwork = getNetwork('testnet')
+
+  public withJellyfishNetwork (networkName: NetworkName): this {
+    this.networkConfig = getNetwork(networkName)
+    return this
+  }
+
+  protected cachedRpcUrl?: string
+
+  /**
+   * Convenience Cmd builder with StartOptions
+   */
+  protected getCmd (): string[] {
+    const { user, password } = this.startOptions
+
+    return [
+      'defid',
+      '-printtoconsole',
+      '-rpcallowip=0.0.0.0/0',
+      '-rpcbind=0.0.0.0',
+      '-rpcworkqueue=512',
+      `-rpcuser=${user}`,
+      `-rpcpassword=${password}`
+    ]
+  }
+
+  /* eslint-disable @typescript-eslint/no-non-null-assertion, no-void */
+  public static readonly PREFIX = 'defichain-testcontainers-'
+
+  /**
+   * Create container and start it immediately waiting for NativeChain to be ready
+   */
+  public async start (): Promise<StartedNativeChainContainer> {
+    const network = await new Network().start()
+
+    const rand = Math.floor(Math.random() * 10000000)
+
+    this.withExposedPorts(...Object.values(this.networkConfig.ports))
+      .withName(`${NativeChainContainer.PREFIX}-${this.networkConfig.name}-${rand}`)
+      .withNetworkMode(network.getName())
+      .withCmd(this.getCmd())
+      .withStartupTimeout(120_000)
```

Comment:
I would move this to the constructor, given this `start ()` design it not Fluent like. When consumer uses the `with...` commands, they expected whatever they injecting in to be replaced. This design will override the replacement when they start the container.

```ts
const started = new NativeChainContainer()
  .withStartupTimeout(150_000)
  .start()
```

The above code expects the timeout to be 120s, however, start will always override the timeout making the 150s injection meaningless.

Reply:
Yup, I will change this.

Worth noting that `MetaChainContainer` also implements it in this manner  https://github.com/DeFiCh/metachain/blob/main/packages/testcontainers/src/MetaChainContainer.ts#L44.

Reply:
Yup, I saw it there too. I've opened an issue there too https://github.com/DeFiCh/metachain/issues/107